File: librarys/med-common-lib/LibWidget/src/main/java/com/medlinker/widget/loading/FrameAnimDrawable.java
package com.medlinker.widget.loading;

import android.animation.ValueAnimator;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.ColorFilter;
import android.graphics.Paint;
import android.graphics.PixelFormat;
import android.graphics.drawable.Animatable;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.view.animation.LinearInterpolator;

import androidx.annotation.IntRange;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**
 * 帧动画Drawable。避免内存泄漏。
 * Created by jiantao on 2017/7/30.
 */

public class FrameAnimDrawable extends Drawable implements Animatable {
    private Resources mResources;
    private int fps = 25;
    private Paint mPaint = null;
    private int[] RES_IDS = null;
    private String[] RES_PATHS = null;
    private int resIndex;

    private int mRepeatCount = ValueAnimator.INFINITE; //播放次数，默认为循环播放

    private ValueAnimator mAnimator;
    private ValueAnimator.AnimatorUpdateListener mAnimUpdateListener;
    private ValueAnimator.AnimatorListener mAnimListener;

    //取第一帧，用于获取图片宽高
    private Drawable mFirstDrawable;
    private int mLength;//资源或文件数量
    private int mBgColor = Color.TRANSPARENT;

    public FrameAnimDrawable() {
    }

    public FrameAnimDrawable(int fps, @NonNull int[] RES_IDS, @NonNull Resources resources) {
        initById(fps, RES_IDS, resources);
    }

    public FrameAnimDrawable(int fps, @NonNull String[] RES_PATHS, @NonNull Resources resources) {
        initByPath(fps, RES_PATHS, resources);
    }

    public void initById(int fps, @NonNull int[] RES_IDS, @NonNull Resources resources) {
        this.fps = fps;
        this.RES_IDS = RES_IDS;
        this.mResources = resources;

        mPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        mPaint.setFilterBitmap(true);
        mPaint.setDither(true);

        if (RES_IDS.length <= 0) {
            throw new RuntimeException(" FrameAnimDrawable RES_IDS can not empty !!!");
        }
        mLength = RES_IDS.length;
        mFirstDrawable = resources.getDrawable(RES_IDS[0]);
        createAnimator();
    }

    public void initByPath(int fps, @NonNull String[] RES_PATHS, @NonNull Resources resources) {
        this.fps = fps;
        this.RES_PATHS = RES_PATHS;
        this.mResources = resources;

        mPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        mPaint.setFilterBitmap(true);
        mPaint.setDither(true);

        if (RES_PATHS.length <= 0) {
            throw new RuntimeException(" FrameAnimDrawable RES_PATHS can not empty !!!");
        }
        mLength = RES_PATHS.length;
        mFirstDrawable = getDrawableFromFile(RES_PATHS[0]);
        createAnimator();
    }

    /**
     * @param path
     * @return
     */
    private BitmapDrawable getDrawableFromFile(String path) {
        Bitmap bitmap = BitmapFactory.decodeFile(path);
        if (bitmap == null) return null;
        return new BitmapDrawable(mResources, bitmap);
    }

    private void createAnimator() {
        mAnimator = ValueAnimator.ofInt(mLength - 1);
        mAnimator.setInterpolator(new LinearInterpolator());
        mAnimator.setRepeatCount(mRepeatCount);
        mAnimator.setRepeatMode(ValueAnimator.RESTART);
        mAnimator.setDuration(mLength / fps * 1000);

        mAnimUpdateListener = new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator animation) {
                invalidate(((int) animation.getAnimatedValue()));
            }
        };
    }

    /**
     * 重绘
     */
    public void invalidate(int index) {
        this.resIndex = index;
        invalidateSelf();
    }

    @Override
    public void draw(@NonNull Canvas canvas) {
        canvas.drawColor(mBgColor);
        if (mPaint != null && mResources != null) {
            if (RES_IDS != null) {
                BitmapDrawable drawable = (BitmapDrawable) mResources.getDrawable(RES_IDS[resIndex % RES_IDS.length]);
                Bitmap bitmap = drawable.getBitmap();
                canvas.drawBitmap(bitmap, 0, 0, mPaint);
            } else if (RES_PATHS != null) {
                Bitmap bitmap = BitmapFactory.decodeFile(RES_PATHS[resIndex % RES_PATHS.length]);
                if (bitmap != null) {
                    canvas.drawBitmap(bitmap, 0, 0, mPaint);
                }
            }
        }
    }

    @Override
    public void setAlpha(@IntRange(from = 0, to = 255) int alpha) {

    }

    @Override
    public void setColorFilter(@Nullable ColorFilter colorFilter) {
        mPaint.setColorFilter(colorFilter);
    }

    @Override
    public int getOpacity() {
        return PixelFormat.OPAQUE;
    }

    @Override
    public void start() {
        // If the animators has not ended, do nothing.
        if (mAnimator == null || mAnimator.isStarted()) return;
        startAnimator();
        invalidateSelf();
    }

    private void startAnimator() {
        mAnimator.addUpdateListener(mAnimUpdateListener);
        if (mAnimListener != null) {
            mAnimator.addListener(mAnimListener);
        }
        mAnimator.start();
    }

    @Override
    public void stop() {
        if (mAnimator != null && mAnimator.isStarted()) {
            mAnimator.removeAllUpdateListeners();
            mAnimator.removeAllListeners();
            mAnimator.end();
        }
    }

    @Override
    public boolean isRunning() {
        return mAnimator.isRunning();
    }

    @Override
    public int getIntrinsicWidth() {
        if (mFirstDrawable != null) {
            return mFirstDrawable.getIntrinsicWidth();
        } else {
            return super.getIntrinsicWidth();
        }
    }

    @Override
    public int getIntrinsicHeight() {
        if (mFirstDrawable != null) {
            return mFirstDrawable.getIntrinsicHeight();
        } else {
            return super.getIntrinsicHeight();
        }
    }

    /**
     * @return 帧数量
     */
    public int getFrameCount() {
        return mLength;
    }

    /**
     * 设置一次动画执行时间
     */
    public void setDuration(float duration) {
        if (mAnimator != null) {
            mAnimator.setDuration((long) (duration * 1000));
        }
    }

    public void setRepeatCount(int count) {
        if (count <= 0) {
            mRepeatCount = ValueAnimator.INFINITE;
        } else {
            mRepeatCount = count - 1;
        }
        if (mAnimator != null) {
            mAnimator.setRepeatCount(mRepeatCount);
            mAnimator.setRepeatMode(ValueAnimator.RESTART);
        }
    }

    public void setAnimListener(ValueAnimator.AnimatorListener listener) {
        mAnimListener = listener;
    }

    public void setBackgroundColor(int color) {
        mBgColor = color;
    }

}

File: librarys/lib-smart-refresh-medlinker-header/LibSmartRefreshHeaderMedlinker/src/main/java/com/medlinker/widget/smart/refresh/header/medlinker/FrameAnimDrawable.kt
package com.medlinker.widget.smart.refresh.header.medlinker

import android.animation.Animator
import android.animation.ValueAnimator
import android.animation.ValueAnimator.AnimatorUpdateListener
import android.content.res.Resources
import android.graphics.*
import android.graphics.drawable.Animatable
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.view.animation.LinearInterpolator
import androidx.annotation.IntRange


/**
 * @author hmy
 * @time 3/23/21 11:51
 */
class FrameAnimDrawable : Drawable, Animatable {
    private var mResources: Resources? = null
    private var fps = 25
    private var mPaint: Paint? = null
    private var RES_IDS: IntArray? = null
    private var RES_PATHS: Array<String?>? = null
    private var resIndex = 0
    private var mRepeatCount = ValueAnimator.INFINITE //播放次数，默认为循环播放
    private var mAnimator: ValueAnimator? = null
    private var mAnimUpdateListener: AnimatorUpdateListener? = null
    private var mAnimListener: Animator.AnimatorListener? = null

    //取第一帧，用于获取图片宽高
    private var mFirstDrawable: Drawable? = null
    private var mLength //资源或文件数量
            = 0
    private var mBgColor = Color.TRANSPARENT

    constructor() {}
    constructor(fps: Int, RES_IDS: IntArray, resources: Resources) {
        initById(fps, RES_IDS, resources)
    }

    constructor(
        fps: Int,
        RES_PATHS: Array<String?>,
        resources: Resources
    ) {
        initByPath(fps, RES_PATHS, resources)
    }

    fun initById(
        fps: Int,
        RES_IDS: IntArray,
        resources: Resources
    ) {
        this.fps = fps
        this.RES_IDS = RES_IDS
        mResources = resources
        mPaint = Paint(Paint.ANTI_ALIAS_FLAG)
        mPaint!!.isFilterBitmap = true
        mPaint!!.isDither = true
        if (RES_IDS.size <= 0) {
            throw RuntimeException(" FrameAnimDrawable RES_IDS can not empty !!!")
        }
        mLength = RES_IDS.size
        mFirstDrawable = resources.getDrawable(RES_IDS[0])
        createAnimator()
    }

    fun initByPath(
        fps: Int,
        RES_PATHS: Array<String?>,
        resources: Resources
    ) {
        this.fps = fps
        this.RES_PATHS = RES_PATHS
        mResources = resources
        mPaint = Paint(Paint.ANTI_ALIAS_FLAG)
        mPaint!!.isFilterBitmap = true
        mPaint!!.isDither = true
        if (RES_PATHS.size <= 0) {
            throw RuntimeException(" FrameAnimDrawable RES_PATHS can not empty !!!")
        }
        mLength = RES_PATHS.size
        mFirstDrawable = getDrawableFromFile(RES_PATHS[0])
        createAnimator()
    }

    /**
     * @param path
     * @return
     */
    private fun getDrawableFromFile(path: String?): BitmapDrawable? {
        val bitmap = BitmapFactory.decodeFile(path) ?: return null
        return BitmapDrawable(mResources, bitmap)
    }

    private fun createAnimator() {
        mAnimator = ValueAnimator.ofInt(mLength - 1)
        mAnimator?.interpolator = LinearInterpolator()
        mAnimator?.repeatCount = mRepeatCount
        mAnimator?.repeatMode = ValueAnimator.RESTART
        mAnimator?.duration = mLength / fps * 1000.toLong()
        mAnimUpdateListener =
            AnimatorUpdateListener { animation -> invalidate(animation.animatedValue as Int) }
    }

    /**
     * 重绘
     */
    fun invalidate(index: Int) {
        resIndex = index
        invalidateSelf()
    }

    override fun draw(canvas: Canvas) {
        canvas.drawColor(mBgColor)
        if (mPaint != null && mResources != null) {
            if (RES_IDS != null) {
                val drawable =
                    mResources!!.getDrawable(RES_IDS!![resIndex % RES_IDS!!.size]) as BitmapDrawable
                val bitmap = drawable.bitmap
                canvas.drawBitmap(bitmap, 0f, 0f, mPaint)
            } else if (RES_PATHS != null) {
                val bitmap =
                    BitmapFactory.decodeFile(RES_PATHS!![resIndex % RES_PATHS!!.size])
                if (bitmap != null) {
                    canvas.drawBitmap(bitmap, 0f, 0f, mPaint)
                }
            }
        }
    }

    override fun setAlpha(
        @IntRange(
            from = 0,
            to = 255
        ) alpha: Int
    ) {
    }

    override fun setColorFilter(colorFilter: ColorFilter?) {
        mPaint!!.colorFilter = colorFilter
    }

    override fun getOpacity(): Int {
        return PixelFormat.OPAQUE
    }

    override fun start() {
        // If the animators has not ended, do nothing.
        if (mAnimator == null || mAnimator!!.isStarted) return
        startAnimator()
        invalidateSelf()
    }

    private fun startAnimator() {
        mAnimator!!.addUpdateListener(mAnimUpdateListener)
        if (mAnimListener != null) {
            mAnimator!!.addListener(mAnimListener)
        }
        mAnimator!!.start()
    }

    override fun stop() {
        if (mAnimator != null && mAnimator!!.isStarted) {
            mAnimator!!.removeAllUpdateListeners()
            mAnimator!!.removeAllListeners()
            mAnimator!!.end()
        }
    }

    override fun isRunning(): Boolean {
        return mAnimator!!.isRunning
    }

    override fun getIntrinsicWidth(): Int {
        return if (mFirstDrawable != null) {
            mFirstDrawable!!.intrinsicWidth
        } else {
            super.getIntrinsicWidth()
        }
    }

    override fun getIntrinsicHeight(): Int {
        return if (mFirstDrawable != null) {
            mFirstDrawable!!.intrinsicHeight
        } else {
            super.getIntrinsicHeight()
        }
    }

    /**
     * @return 帧数量
     */
    fun getFrameCount(): Int {
        return mLength
    }

    /**
     * 设置一次动画执行时间
     */
    fun setDuration(duration: Float) {
        if (mAnimator != null) {
            mAnimator!!.duration = (duration * 1000).toLong()
        }
    }

    fun setRepeatCount(count: Int) {
        mRepeatCount = if (count <= 0) {
            ValueAnimator.INFINITE
        } else {
            count - 1
        }
        if (mAnimator != null) {
            mAnimator!!.repeatCount = mRepeatCount
            mAnimator!!.repeatMode = ValueAnimator.RESTART
        }
    }

    fun setAnimListener(listener: Animator.AnimatorListener?) {
        mAnimListener = listener
    }

    fun setBackgroundColor(color: Int) {
        mBgColor = color
    }
}
